unsupervised_da_exp.py

```diff
@@ -0,0 +1,175 @@
+                      default="discr",
+                      help="domain adaptation type"
+                      "(discr = discriminative"
+                      "mm = moment matching"
+                      "mmv2 = moment matching with discrepancy")
+  parser.add_argument("--da_loss_weight",
+                      type=float,
+                      default=1.,
+                      help="DA loss weight.")
+  return parser
+
+def get_da_model(classifier, num_layers,
+                 ds_list, target_domain,
+                 losses, loss_weights,
+                 da_mode="discr", block_shape=(16, 16, 40),
+                 da_loss_weight=1.,
+                 lrate=0.002):
+  """Returns a compiled DA model of a given type.
+
+  Arguments:
+  classifier: source classifier model.
+  num_layers: number of residual blocks.
+  ds_list: list of dataset names.
+  target_domain: target domain name.
+  losses: dictionary of classification losses for all domains.
+  loss_weights: dictionary of classification loss weights.
+  da_mode: domain adaptation mode: discr, mm or mmv2.
+  block_shape: residual block output shape.
+  da_loss_weight: DA loss weight.
+  """
+  cl_optimizer = tfk.optimizers.RMSprop(learning_rate=lrate)
+# creating a DA model
+  if "discr" in da_mode:
+    discriminator = adv.get_discriminator(
+        input_shape=block_shape, num_layers=2,
+        dropout=0, activation=tf.nn.leaky_relu)
+
+    da_model = adv.AdversarialClassifier(
+        discriminator=discriminator, classifier=classifier,
+        domains=ds_list, target_domain=target_domain,
+        num_layers=num_layers)
+    da_model.compile(
+        cl_optimizer=cl_optimizer,
+        d_optimizer=tfk.optimizers.SGD(learning_rate=0.0002),
+        adv_optimizer=tfk.optimizers.Adam(learning_rate=0.0002),
+        loss_fn=tfk.losses.BinaryCrossentropy(from_logits=False),
+        cl_losses=losses,
+        cl_loss_weights=loss_weights,
+        d_loss_weight=1.,
+        adv_loss_weight=da_loss_weight)
+
+  elif "mmv2" in da_mode:
+    raise NotImplementedError("mmv2 is not implemented yet.")
+  else:
+    da_model = mm.MomentMatchingClassifier(
+        classifier=classifier,
+        domains=ds_list, target_domain=target_domain,
+        num_layers=num_layers)
+    da_model.compile(
+        cl_optimizer=cl_optimizer,
+        mm_optimizer=tfk.optimizers.Adam(learning_rate=0.001),
+        mm_loss_weight=1,
+        cl_losses=losses,
+        cl_loss_weights=loss_weights)
+  return da_model
+
+
+
+def main():
+  parser = get_custom_parser()
+  args = parser.parse_args()
+  args.save_path = os.path.abspath(args.save_path)
+  exp_path = os.path.join(args.save_path, args.name)
+  if not os.path.exists(exp_path):
+    os.makedirs(exp_path)
+  args_util.write_config(args, exp_path)
+  results_file = os.path.join(exp_path, "results.txt")
+
+  if len(args.ckpt_path) == 0:
+    args.ckpt_path = None
+  # obtaining the datasets
```

Comment:
This is a clear example where a single function call to `prepare_datasets` (or something like this) would allow to: (a) clean the code and isolate a semantic code element, (b) avoid the need for the comment altogether.